cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java

```diff
@@ -195,8 +195,21 @@ public void onChange(Collection<PartitionReplica> partitionReplicas) {
         @Override
         public void finished(Throwable failureCause) {
           if (service != null) {
-            service.stopAndWait();
-            service = null;
+            try {
+              service.stopAndWait();
+              service = null;
+            } finally {
+              // we need to set completion status for the service, if failure case is non-null set as exception.
+              if (failureCause != null) {
+                LOG.error("No more changes will be notified to the service: {}.", serviceName, failureCause);
+                completion.setException(failureCause);
+              } else {
+                LOG.info("No more changes will be notified to the service: {}.", serviceName);
+                completion.set(null);
+              }
+            }
+          } else {
+            completion.set(null);
```

Comment:
yeah, you are right, it seems like this completion future variable is not actually needed, based on the logic in ResourceHandler, we would want to restart metrics/log.saver services on partition change's and we would not want to stop the resource balancer service on those scenarios as well as  exception during starting/stopping metrics/log.saver service. Its not clear to me when the ``ResourceBalancer#shutdown`` would be called. 
But i looked at the logs again for the issue, and it's clear the ``onChange`` was called twice(both happened on the same timestamp), first time with non-empty partitions list and the second time with an empty list, the first call started the metrics.processor service, while the second call stopped the service, which is expected as we stop the service during ``onChange`` and don't start it back if the list is empty.  However its not clear why an empty partition's list  was sent..